File: app/src/main/java/de/xikolo/controllers/helper/SectionDownloadHelper.java
package de.xikolo.controllers.helper;

import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Observer;

import de.xikolo.controllers.dialogs.MobileDownloadDialog;
import de.xikolo.controllers.dialogs.ModuleDownloadDialog;
import de.xikolo.controllers.dialogs.ModuleDownloadDialogAutoBundle;
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminate;
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminateAutoBundle;
import de.xikolo.managers.DownloadManager;
import de.xikolo.models.Course;
import de.xikolo.models.DownloadAsset;
import de.xikolo.models.Item;
import de.xikolo.models.Section;
import de.xikolo.models.dao.VideoDao;
import de.xikolo.network.jobs.ListItemsWithContentForSectionJob;
import de.xikolo.network.jobs.base.NetworkCode;
import de.xikolo.network.jobs.base.NetworkState;
import de.xikolo.network.jobs.base.NetworkStateLiveData;
import de.xikolo.storages.ApplicationPreferences;
import de.xikolo.utils.LanalyticsUtil;
import de.xikolo.utils.NetworkUtil;

public class SectionDownloadHelper {

    private FragmentActivity activity;

    private boolean sdVideo;

    private boolean hdVideo;

    private boolean slides;

    private DownloadManager downloadManager;

    public SectionDownloadHelper(FragmentActivity activity) {
        this.activity = activity;
        this.downloadManager = new DownloadManager(activity);
    }

    public void initSectionDownloads(final Course course, final Section section) {
        ModuleDownloadDialog listDialog = ModuleDownloadDialogAutoBundle.builder(section.title).build();
        listDialog.setListener((dialog, hdVideo, sdVideo, slides) -> {
            SectionDownloadHelper.this.hdVideo = hdVideo;
            SectionDownloadHelper.this.sdVideo = sdVideo;
            SectionDownloadHelper.this.slides = slides;

            final ApplicationPreferences appPreferences = new ApplicationPreferences();

            if (hdVideo || sdVideo || slides) {
                if (NetworkUtil.isOnline()) {
                    if (NetworkUtil.getConnectivityStatus() == NetworkUtil.TYPE_MOBILE &&
                            appPreferences.isDownloadNetworkLimitedOnMobile()) {
                        MobileDownloadDialog permissionDialog = new MobileDownloadDialog();
                        permissionDialog.setListener(dialog1 -> {
                            appPreferences.setDownloadNetworkLimitedOnMobile(false);
                            startSectionDownloads(course, section);
                        });
                        permissionDialog.show(activity.getSupportFragmentManager(), MobileDownloadDialog.TAG);
                    } else {
                        startSectionDownloads(course, section);
                    }
                } else {
                    NetworkUtil.showNoConnectionToast();
                }
            }
        });
        listDialog.show(activity.getSupportFragmentManager(), ModuleDownloadDialog.TAG);
    }

    private void startSectionDownloads(final Course course, final Section section) {
        LanalyticsUtil.trackDownloadedSection(section.id, course.id, hdVideo, sdVideo, slides);

        final ProgressDialogIndeterminate dialog = ProgressDialogIndeterminateAutoBundle.builder().build();
        dialog.show(activity.getSupportFragmentManager(), ProgressDialogIndeterminate.TAG);

        NetworkStateLiveData itemRequestNetworkState = new NetworkStateLiveData();
        itemRequestNetworkState.observeForever(new Observer<NetworkState>() {
            @Override
            public void onChanged(NetworkState networkState) {
                if (networkState.getCode() == NetworkCode.STARTED) {
                    return;
                }

                if (networkState.getCode() == NetworkCode.SUCCESS) {
                    dialog.dismissAllowingStateLoss();
                    for (Item item : section.getAccessibleItems()) {
                        if (item.contentType.equals(Item.TYPE_VIDEO)) {
                            if (sdVideo) {
                                startDownload(new DownloadAsset.Course.Item.VideoSD(item, VideoDao.Unmanaged.find(item.contentId)));
                            }
                            if (hdVideo) {
                                startDownload(new DownloadAsset.Course.Item.VideoHD(item, VideoDao.Unmanaged.find(item.contentId)));
                            }
                            if (slides) {
                                startDownload(new DownloadAsset.Course.Item.Slides(item, VideoDao.Unmanaged.find(item.contentId)));
                            }
                        }
                    }
                } else {
                    dialog.dismiss();
                }

                itemRequestNetworkState.removeObserver(this);
            }
        });
        new ListItemsWithContentForSectionJob(section.id, itemRequestNetworkState, false).run();
    }

    private void startDownload(DownloadAsset.Course.Item item) {
        if (!downloadManager.downloadExists(item)
            && !downloadManager.downloadRunning(item)) {
            downloadManager.startAssetDownload(item);
        }
    }

}

File: app/src/main/java/de/xikolo/controllers/helper/SectionDownloadHelper.kt
package de.xikolo.controllers.helper

import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Observer
import de.xikolo.R
import de.xikolo.controllers.dialogs.MobileDownloadDialog
import de.xikolo.controllers.dialogs.ModuleDownloadDialog
import de.xikolo.controllers.dialogs.ModuleDownloadDialogAutoBundle
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminate
import de.xikolo.controllers.dialogs.ProgressDialogIndeterminateAutoBundle
import de.xikolo.download.DownloadItem
import de.xikolo.download.DownloadStatus
import de.xikolo.models.Course
import de.xikolo.models.DownloadAsset
import de.xikolo.models.Item
import de.xikolo.models.Section
import de.xikolo.models.dao.VideoDao
import de.xikolo.network.jobs.ListItemsWithContentForSectionJob
import de.xikolo.network.jobs.base.NetworkCode
import de.xikolo.network.jobs.base.NetworkState
import de.xikolo.network.jobs.base.NetworkStateLiveData
import de.xikolo.storages.ApplicationPreferences
import de.xikolo.utils.LanalyticsUtil
import de.xikolo.utils.extensions.ConnectivityType
import de.xikolo.utils.extensions.connectivityType
import de.xikolo.utils.extensions.isOnline
import de.xikolo.utils.extensions.showToast

class SectionDownloadHelper(private val activity: FragmentActivity) {

    fun initSectionDownloads(course: Course, section: Section) {
        val listDialog = ModuleDownloadDialogAutoBundle.builder(section.title).build()

        listDialog.listener = object : ModuleDownloadDialog.ItemSelectionListener {

            override fun onSelected(dialog: DialogFragment, video: Boolean, slides: Boolean) {
                val appPreferences = ApplicationPreferences()
                if (video || slides) {
                    if (activity.isOnline) {
                        if (activity.connectivityType === ConnectivityType.CELLULAR && appPreferences.isDownloadNetworkLimitedOnMobile) {
                            val permissionDialog = MobileDownloadDialog()

                            permissionDialog.listener =
                                object : MobileDownloadDialog.MobileDownloadGrantedListener {

                                    override fun onGranted(dialog: DialogFragment) {
                                        appPreferences.isDownloadNetworkLimitedOnMobile = false
                                        startSectionDownloads(course, section, video, slides)
                                    }
                                }
                            permissionDialog.show(
                                activity.supportFragmentManager,
                                MobileDownloadDialog.TAG
                            )
                        } else {
                            startSectionDownloads(course, section, video, slides)
                        }
                    } else {
                        activity.showToast(R.string.toast_no_network)
                    }
                }
            }
        }

        listDialog.show(activity.supportFragmentManager, ModuleDownloadDialog.TAG)
    }

    private fun startSectionDownloads(
        course: Course,
        section: Section,
        video: Boolean,
        slides: Boolean
    ) {
        LanalyticsUtil.trackDownloadedSection(section.id, course.id, video, false, slides)

        val dialog = ProgressDialogIndeterminateAutoBundle.builder().build()
        dialog.show(activity.supportFragmentManager, ProgressDialogIndeterminate.TAG)

        val itemRequestNetworkState = NetworkStateLiveData()
        itemRequestNetworkState.observeForever(object : Observer<NetworkState> {
            override fun onChanged(networkState: NetworkState) {
                if (networkState.code === NetworkCode.STARTED) {
                    return
                }

                if (networkState.code === NetworkCode.SUCCESS) {
                    dialog.dismissAllowingStateLoss()
                    for (item in section.accessibleItems) {
                        if (item.contentType == Item.TYPE_VIDEO) {
                            if (video) {
                                startDownload(
                                    DownloadAsset.Course.Item.VideoHLS(
                                        item,
                                        VideoDao.Unmanaged.find(item.contentId)!!,
                                        ApplicationPreferences().videoDownloadQuality
                                    )
                                )
                            }

                            if (slides) {
                                startDownload(
                                    DownloadAsset.Course.Item.Slides(
                                        item,
                                        VideoDao.Unmanaged.find(item.contentId)!!
                                    )
                                )
                            }
                        }
                    }
                } else {
                    dialog.dismiss()
                }

                itemRequestNetworkState.removeObserver(this)
            }
        })
        ListItemsWithContentForSectionJob(section.id, itemRequestNetworkState, false).run()
    }

    private fun startDownload(item: DownloadItem<*, *>) {
        if (item.status.state == DownloadStatus.State.DELETED) {
            item.start(activity)
        }
    }
}
